src/common/ast-path.js

```diff
@@ -89,24 +115,55 @@ class AstPath {
       stack.push(name, value);
     }
 
-    for (let i = 0; i < value.length; ++i) {
-      stack.push(i, value[i]);
-      callback(this, i, value);
-      stack.length -= 2;
-    }
-
-    stack.length = length;
+    const iteratee = (index) =>
+      tryFinally(
+        () => {
+          stack.push(index, value[index]);
+          return callback(this, index, value);
+        },
+        () => {
+          stack.length -= 2;
+        }
+      );
+
+    return tryFinally(
+      () => {
+        let promise;
+
+        for (let index = 0; index < value.length; index++) {
+          if (promise) {
+            promise = promise.then(() => iteratee(index));
+            continue;
+          }
+
+          const result = iteratee(index);
+
+          if (isThenable(result)) {
+            promise = result;
+          }
+        }
+
+        return promise;
+      },
+      () => {
+        stack.length = length;
+      }
+    );
   }
 
   // Similar to AstPath.prototype.each, except that the results of the
   // callback function invocations are stored in an array and returned at
   // the end of the iteration.
   map(callback, ...names) {
-    const result = [];
-    this.each((path, index, value) => {
-      result[index] = callback(path, index, value);
+    const array = [];
```

Comment:
More meaningful name?

```suggestion
    const results = [];
```

Reply:
https://github.com/prettier/prettier/pull/13104/commits/b883bf86c64f35cc2d38493a1d8e7522d83f63b4